tree_arena/src/tree_arena_unsafe.rs

```diff
@@ -0,0 +1,617 @@
+    parent_arena: &'arena DataMap<T>,
+    /// The parent id for these children
+    id: Option<NodeId>,
+    /// Array of children
+    #[expect(dead_code, reason = "Kept for possible future use")]
+    child_arr: &'arena [NodeId],
+}
+
+impl<Item> Clone for ArenaRef<'_, Item> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<Item> Copy for ArenaRef<'_, Item> {}
+
+impl<T> Clone for ArenaRefChildren<'_, T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<Item> Copy for ArenaRefChildren<'_, Item> {}
+
+/// A reference type giving mutable access to an arena item and its children.
+///
+/// When you borrow an item from a [`TreeArena`], it returns an `ArenaMut`.
+/// This struct holds three fields:
+///  - the id of its parent.
+///  - a reference to the item itself.
+///  - an [`ArenaMutChildren`] handle to access its children.
+///
+/// Because the latter two are disjoint references, you can mutate the node's value
+/// and its children independently without invalidating the references.
+///
+/// You can iterate over its children to get access to child `ArenaMut` handles.
+#[derive(Debug)]
+pub struct ArenaMut<'arena, T> {
+    /// Parent of the Node
+    pub parent_id: Option<NodeId>,
+    /// Item in the node
+    pub item: &'arena mut T,
+    /// Children of the node
+    pub children: ArenaMutChildren<'arena, T>,
+}
+
+/// A handle giving mutable access to an arena item's children.
+///
+/// See [`ArenaMut`] for more information.
+///
+/// This stores all the permissions for what nodes can be accessed from the current node
+/// As such if a [`std::mem::swap`] is used to swap the children of two trees,
+/// each tree will still have the correct permissions. This also stores the roots, and so
+/// that will also be in a consistent state
+#[derive(Debug)]
+pub struct ArenaMutChildren<'arena, T> {
+    /// The associated data arena
+    parent_arena: &'arena mut DataMap<T>,
+    /// The parent id for these children
+    id: Option<NodeId>,
+    /// Array of children
+    child_arr: &'arena mut Vec<NodeId>,
+}
+
+impl<T> DataMap<T> {
+    fn new() -> Self {
+        Self {
+            items: HashMap::new(),
+            parents: HashMap::new(),
+            roots: Box::new(UnsafeCell::new(Vec::new())),
+        }
+    }
+
+    /// Find an item in the tree.
+    ///
+    /// Returns a shared reference to the item if present.
+    ///
+    /// Time Complexity O(1)
+    fn find(&self, id: impl Into<NodeId>) -> Option<ArenaRef<'_, T>> {
+        // SAFETY
+        // We need there to be no mutable access to the node
+        // Mutable access to the node would imply there is some &mut self
+        // As we are taking &self, there can be no mutable access to the node
+        // Thus this is safe
+        let id: NodeId = id.into();
+        let parent_id = *self.parents.get(&id)?;
+        self.items
+            .get(&id)
+            .and_then(|item| unsafe { item.get().as_ref() })
```

Comment:
I think unwrap is definitely less preferable (makes more sense to return `None` in the case of the `NodeId` not being in the tree rather than panic). Would you say breaking it down as 

```rust
let node_cell = self.items.get(&id)?;

let TreeNode {
    item,
    children: child_arr,
} = unsafe { node_cell.get().as_ref()? };

let children = ArenaRefChildren {
    parent_arena: self,
    id: Some(id),
    child_arr,
};
Some(ArenaRef {
    parent_id,
    item,
    children,
})
  ```
  is preferable

Reply:
I meant unwrap for `get().as_ref()`, since `UnsafeCell::get()` will never return a null pointer, but either way `?` works. The branches should get elided by the compiler anyway.

> Would you say breaking it down as [...] is preferable?

I'm not a fan of destructuring, but yes, this is much better.